### ISSUE TITLE
HTTP Server connection are being closed before use

### ISSUE THREAD

**Issue by inkusgames** [deep-entertainment/issues#52]
# Project: GodotTPD

- [ ] Safehouse
- [ ] sheets to frames addon
- [x] Godot asset lib Github action

**Describe the bug**
When using godottpd client connections are sometimes flushed before responses are given. 

**To Reproduce**
1. Host a site with multiple files and scripts that download them to create a lot of traffic.
2. Refresh that site you sill see some requests die without response

**Expected behavior**
All requests should be serviced.

**Versions**
[6825e55](https://github.com/deep-entertainment/godottpd/commit/6825e55b72831fd9a3ace1a49af7b00280404460)

**Additional context**
I managed to solve the issue with the following changes to http_server.gd

``` gd

# Handle possibly incoming requests
func _process(_delta: float) -> void:
	if _server:
		while _server.is_connection_available():
			var new_client = _server.take_connection()
			if new_client:
				self._clients.append(new_client)
		for client in self._clients:
			client.poll()
			if client.get_status() == StreamPeerTCP.STATUS_CONNECTED:
				var bytes = client.get_available_bytes()
				if bytes > 0:
					var request_string = client.get_utf8_string(bytes)
					self._handle_request(client, request_string)
		_remove_disconnected_clients()


func _remove_disconnected_clients():
	self._clients = self._clients.filter(func(c: StreamPeerTCP): return c.get_status() == StreamPeerTCP.STATUS_CONNECTED or c.get_status() == StreamPeerTCP.STATUS_CONNECTING)

```

The primary fix is in _remove_disconnected_clients() it did not have the return and adding the status of StreamPeerTCP.STATUS_CONNECTING to conditions to preserve the connection.

Additionally adding new clients is done in a while loop as apposed to one per frame.

Lastly adding client.poll() before testing the status of the connection to comply with the documented behavior of StreamPeerTCP

**Comment by dploeger** [comment]
Thanks a lot. This is a great idea. I just don't understand why clients in CONNECTING state should be removed  in _remove_disconnected_clients. Could you please elaborate on that? Thanks

**Comment by inkusgames** [comment]
So the way it works is a filter that removes clients. 

A new list of clients is set using a filter that returns true for each elements you want to keep.

This removes any that don't match that filter if statement. The `if` is to determine which clients to keep.

The current code is

``` gdscript

func _remove_disconnected_clients():
	self._clients = self._clients.filter(func(c: StreamPeerTCP): c.get_status() == StreamPeerTCP.STATUS_CONNECTED)

```

So the filter was **not returning true or false** so it was removing all clients from the list as none would match the criteria. I also added the status of `STATUS_CONNECTING` as a valid reason to keep a connection.

The filter documentation can be read here https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-filter

Changing it to

``` gdscript
func _remove_disconnected_clients():
	self._clients = self._clients.filter(func(c: StreamPeerTCP): return c.get_status() == StreamPeerTCP.STATUS_CONNECTED or c.get_status() == StreamPeerTCP.STATUS_CONNECTING)
```

Correctly returns a value in the filter and doesn't remove clients still connecting.

**Comment by dploeger** [comment]
> A new list of clients is set using a filter that returns true for each elements you want to keep.

Oh right! The filter KEPT connected clients and not remove them. Thanks for explaining my own code. 😅

> So the filter was **not returning true or false** so it was removing all clients from the list as none would match the criteria. I also added the status of `STATUS_CONNECTING` as a valid reason to keep a connection.

I think you're mistaken there. The return keyword isn't required (at least I think it isn't because that's how lambdas in other languages work), so previously it should've kept connected clients. But you're right, connectING clients should be kept as well. 

Thanks!

I'll work that into the code.

**Comment by inkusgames** [comment]
> I think you're mistaken there. The return keyword isn't required (at least I think it isn't because that's how lambdas in other languages work), so previously it should've kept connected clients. But you're right, connectING clients should be kept as well.

I don't think I am mistaken.

This is not a lambda function is just a Callable that should return if the value is to be kept. The godot example code looks as follows.

```gdscript
func is_even(number):
	return number % 2 == 0

func _ready():
	print([1, 4, 5, 8].filter(is_even)) # Prints [4, 8]

	# Same as above, but using a lambda function.
	print([1, 4, 5, 8].filter(func(number): return number % 2 == 0))

```
In both the case of the function being anonymous and just been referenced both return a value.  I am not aware of GDScript treating the last value as a return stack item.

If you exclude the return as follows

```gdscript

func _ready():
	# The below will print an empty array
	print([1, 4, 5, 8].filter(func(number): number % 2 == 0))

```
It will print an empty array

**Comment by dploeger** [comment]
Ohhh, okay. Thanks for clearing this up for me! I'll work it in.

**Comment by inkusgames** [comment]
Pleasure, and thanks for building the Web server. Saved me a lot of time that's for sure.

**Comment by dploeger** [comment]
Fixed in 0.4.2
